diff -ur mod_session-apache2.2-ports.orig/apu/apr_crypto.c mod_session-apache2.2-ports/apu/apr_crypto.c
--- mod_session-apache2.2-ports.orig/apu/apr_crypto.c	2014-08-22 00:29:54.000000000 +0900
+++ mod_session-apache2.2-ports/apu/apr_crypto.c	2015-04-18 00:00:00.000000000 +0900
@@ -26,6 +26,11 @@
 #include "apr_thread_mutex.h"
 #include "apr_lib.h"
 
+
+/* :: include ported headers and routines */
+#include "parts/use_apr_crypto.h"
+
+
 #if APU_HAVE_CRYPTO
 
 #include "apu_internal.h"
diff -ur mod_session-apache2.2-ports.orig/apu/apr_crypto_openssl.c mod_session-apache2.2-ports/apu/apr_crypto_openssl.c
--- mod_session-apache2.2-ports.orig/apu/apr_crypto_openssl.c	2011-12-09 21:05:40.000000000 +0900
+++ mod_session-apache2.2-ports/apu/apr_crypto_openssl.c	2015-04-18 00:00:00.000000000 +0900
@@ -26,6 +26,11 @@
 #include "apr_time.h"
 #include "apr_buckets.h"
 
+
+/* :: include ported headers and routines */
+#include "parts/use_apr_crypto.h"
+
+
 #include "apr_crypto_internal.h"
 
 #if APU_HAVE_CRYPTO
diff -ur mod_session-apache2.2-ports.orig/mod_session.c mod_session-apache2.2-ports/mod_session.c
--- mod_session-apache2.2-ports.orig/mod_session.c	2014-01-24 22:25:27.000000000 +0900
+++ mod_session-apache2.2-ports/mod_session.c	2015-04-18 00:00:00.000000000 +0900
@@ -22,6 +22,13 @@
 #include "http_request.h"
 #include "http_protocol.h"
 
+
+/* :: include ported headers and routines */
+#include "parts/from_httpd.h"
+#include "parts/from_http_log.h"
+#include "parts/from_http_config.h"
+
+
 #define SESSION_EXPIRY "expiry"
 #define HTTP_SESSION "HTTP_SESSION"
 
diff -ur mod_session-apache2.2-ports.orig/mod_session_cookie.c mod_session-apache2.2-ports/mod_session_cookie.c
--- mod_session-apache2.2-ports.orig/mod_session_cookie.c	2013-07-07 22:50:54.000000000 +0900
+++ mod_session-apache2.2-ports/mod_session_cookie.c	2015-04-18 00:00:00.000000000 +0900
@@ -20,6 +20,13 @@
 #include "http_log.h"
 #include "util_cookies.h"
 
+
+/* :: include ported headers and routines */
+#include "parts/from_http_log.h"
+#include "parts/from_http_config.h"
+#include "util_cookies.c"
+
+
 #define MOD_SESSION_COOKIE "mod_session_cookie"
 
 module AP_MODULE_DECLARE_DATA session_cookie_module;
diff -ur mod_session-apache2.2-ports.orig/mod_session_crypto.c mod_session-apache2.2-ports/mod_session_crypto.c
--- mod_session-apache2.2-ports.orig/mod_session_crypto.c	2013-12-27 03:08:28.000000000 +0900
+++ mod_session-apache2.2-ports/mod_session_crypto.c	2015-04-18 00:00:00.000000000 +0900
@@ -22,6 +22,15 @@
 #include "http_log.h"
 #include "http_core.h"
 
+
+/* :: include ported headers and routines */
+#include "parts/from_httpd.h"
+#include "parts/from_http_log.h"
+#include "parts/from_http_config.h"
+#include "parts/from_http_core.h"
+#include "parts/use_apr_crypto.h"
+
+
 #if APU_MAJOR_VERSION == 1 && APU_MINOR_VERSION < 4
 
 #error session_crypto_module requires APU v1.4.0 or later
diff -ur mod_session-apache2.2-ports.orig/util_cookies.h mod_session-apache2.2-ports/util_cookies.h
--- mod_session-apache2.2-ports.orig/util_cookies.h	2012-01-09 22:18:26.000000000 +0900
+++ mod_session-apache2.2-ports/util_cookies.h	2015-04-18 00:00:00.000000000 +0900
@@ -22,6 +22,11 @@
 #ifndef UTIL_COOKIES_H
 #define UTIL_COOKIES_H
 
+
+/* :: include ported headers and routines */
+#include "parts/from_ap_config.h"
+
+
 #ifdef __cplusplus
 extern "C" {
 #endif
